src/main/kotlin/org/rust/lang/refactoring/RsImportOptimizer.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.refactoring
+
+import com.intellij.lang.ImportOptimizer
+import com.intellij.psi.PsiFile
+import org.rust.ide.formatter.processors.asTrivial
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.RsMod
+import org.rust.lang.core.psi.ext.childrenOfType
+import org.rust.openapiext.runWriteCommandAction
+
+class RsImportOptimizer : ImportOptimizer {
+    override fun supports(file: PsiFile?): Boolean = file is RsFile
+
+    override fun processFile(file: PsiFile?) = Runnable {
+        OptimizeProcess(file as RsFile).execute()
+    }
+
+    private class OptimizeProcess(private val mod: RsMod) {
+        fun execute() {
+            val uses = mod.childrenOfType<RsUseItem>()
+            if (uses.isEmpty()) {
+                return
+            }
+            mod.project.runWriteCommandAction { replaceOrder(mod, uses) }
+            val mods = mod.childrenOfType<RsMod>()
+            mods.forEach { OptimizeProcess(it).execute() }
+        }
+    }
+
+    companion object {
+
+        private fun optimizeUseSpeck(psiFactory: RsPsiFactory, useSpeck: RsUseSpeck) {
```

Comment:
I think you can easily support `useSpeck` optimizing for each level - just call `optimizeUseSpeck` for every child `useSpeck`

Reply:
done